### PR TITLE
Remove useless contiguous calls from torch.matmul

### PR DESCRIPTION
**Pull request by albanD**
This reduces the memory usage of matmul significantly for expanded batch size.

This reduces the peak memory usage of
```
a = torch.rand(1, 1024, 1024, device="cuda")
b = torch.rand(1024, 1024, 1, device="cuda")

out = torch.matmul(a, b)
```
From 4GB to 16MB which is not too bad.

It also fixes the same problem when `b` is not batched.

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -3432,6 +3432,34 @@ def test_cuda_device_memory_allocated(self):
...
+        def get_max_used():
+            torch.cuda.synchronize()
+            val = torch.cuda.max_memory_allocated()
+            torch.cuda.reset_peak_memory_stats()
+            return val
+
+        a = torch.rand(1, 32, 32, device="cuda")
+        b = torch.rand(24, 32, 1, device="cuda")
+
+        get_max_used()
```
Comment (ngimel) [inline test/test_cuda.py:3445]: The return value of this call is ignored, are you using it to reset_peak_memory only?

Comment (albanD) [reply test/test_cuda.py:3445]: Yes.

**Review by ngimel — APPROVED** [review]
Did you check that this test fails without your changes? The matrices might be small enough that the difference is hidden behind the size of caching allocator blocks.

**Comment by albanD** [comment]
Yes I wrote the test before doing the update and it was failing.

**Pull request closed without merging**